pkg/controllers/node/finalizer.go

```diff
@@ -0,0 +1,38 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha3"
+	"github.com/awslabs/karpenter/pkg/utils/functional"
+	v1 "k8s.io/api/core/v1"
+)
+
+// Finalizer is a tiny reconciler that ensures nodes have the termination
+// finalizer. This protects against instances that launch when Karpenter fails
+// to create the node object. In this case, the node will come online without
+// the termination finalizer. This controller will update the node accordingly.
+type Finalizer struct {}
+
+// Reconcile adds the termination finalizer if the node is not deleting
+func (r *Finalizer) Reconcile(n *v1.Node) error {
+	if !n.DeletionTimestamp.IsZero() {
```

Comment:
It's possible a node could be deleting because it had another deletionTimestamp on it, but not the TerminationFinalizer. In this case, if the other finalizers don't implement instance deletion logic, this might leak.

Reply:
Potentially some daemonsets that only tolerate specific taints, but most tolerate all. We should stay vigilant on this one. Worst case, we alter the behavior do set node.spec.unschedulable, but it's hard to differentiate this against the cordon behavior.